text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+
+To support integrations, we will need to publish a list of all messages.
+We will use static code analysis to automatically compile and publish this list without adding additional maintenance burden on to the team.
+
+#### Synth-time Errors
+
+This includes making changes to all errors thrown from the construct library.
+If an integrator includes the CDK App in the same process, errors can either be handled directly or will be rethrown.
+Cloud Executables that are going through a child process will eventually print all errors to stderr.
+For these we are going to detect errors from the output streams and convert them to JavaScript errors.
+The Toolkit will then throw them as before.
+
+#### Future Extensions
+
+This proposal focuses on a feature set required for GA.
+However the design allows for easy future extensions:
+
+- Additional actions can be migrated to the [Programmatic Toolkit]
+- The `IoHost` can be extended to include other I/O tasks like filesystem interactions.
+  This would allow implementations that are truly, in-memory like in browser environments.
+- The Toolkit setup can allow additional configuration, like providing a custom SDK or authentication configuration.
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
+
+#### SdkProvider
+
+> Designing and committing to a public interface for SDK Provider will require some dedicated design.
+> The recent SDKv3 migration highlighted that the current design needs work.
+> Exposing the full SDK to users also puts a lot of requirements and work on them.
+> Instead we might want to consider a more focused approach like an interface for authentication config.
+> This is still in scope for phase 3, but we will run a separate RFC and design for it.
+
+The main purpose of this feature is to expose SDK configuration to an integrator.
+We will be providing our default implementation as a drop-in solution for CLI-like environments.
+However other possible scenarios like standalone apps or an IDE plugin will require very different setups.
+
+Authentication is the primary focus of this public contract.
+However from our own experience we are aware that authentication can also extend to other parts of SDK configuration.
+For example proxies require HTTP transport configuration and default timeouts and retries are also worth considering.
+
+The existing plugin system is a superset of the SDK v3's `AwsCredentialIdentityProvider` type as it introduces a mode (reading or writing).
+The public contract will subsume and formalize this.
+
+#### Readme
+
+The `AwsSdkProvider` defines how the toolkit interacts with AWS services.
+A default provider is available.
+You can provide a custom class implementing the `IAwsSdkProvider` interface.
+A standard `SdkProviderForCdk` is available that includes all features of the AWS CDK CLI [TODO: this might end up in a separate package].
```

Comment:
I don't follow. Is this part of the previous section? Also it doesn't sound like an alternate solution, but rather a possible future extension?

Reply:
Sure. The currently proposed alternative is to not offer a configuration.